### PR TITLE
Add double version of dataset_as_flex_int

### PR DESCRIPTION
**Pull request by phyy-nx**
This allows reading NeXus data saved with double precision, such as either the JF16M corrected data or the JF16M gain map/pedestals.

Would appreciate a quick check from @graeme-winter ensuring this doesn't break DLS Eiger datasets.  Thanks :)

### CODE REVIEW

File: format/nexus.py [hunk 1]
```diff
@@ -1632,8 +1640,8 @@ def __getitem__(self, index):
         d = self.lookup[index]
         i = index - self.offset[d]
         N, height, width = self.datasets[d].shape
-        data_as_flex = dataset_as_flex_int(
-            self.datasets[d].id.id,
```
Comment (graeme-winter) [inline format/nexus.py]: Is it correct that `.id.id` did nothing? I would find it surprising if @jmp1985 left this in there for no reason. I assume that the code worked, though...

File: format/boost_python/nexus_ext.cc [hunk 2]
```diff
@@ -20,11 +20,34 @@
...
+  inline herr_t custom_read<int>(hid_t dataset_id, hid_t mem_space_id, hid_t file_space_id,
+                          scitbx::af::versa<int, scitbx::af::flex_grid<> > data) {
+    return H5Dread(
+      dataset_id, H5T_NATIVE_INT, mem_space_id, file_space_id, H5P_DEFAULT, &data[0]);
+  }
+
+  /**
+   * Custom read function for doubles
+   */
+  template<>
```
Comment (graeme-winter) [inline format/boost_python/nexus_ext.cc]: If you are adding `double`, is it worth adding `float` at the same time for future proofing?

**Comment by phyy-nx** [comment]
It’s not that it needs big precision, dataset_as_flex_int just rounds the data, if it is already stored as floating point. This happens if a) the corrections were already applied or b) the raw data (stored as int) is being corrected in memory using the gain and the pedestal (stored as floats). It’s possible that float would work instead of double. I’ll check how the data is stored tomorrow.

.id.id is moved to the new function dataset_as_flex at the top of nexus.py

I purport that dataset.dtype doesn’t trigger a read based on my testing. It’s not a numpy object but an h5py object and dtype returns immediately. I can still cache it tho. 

I don’t know if there is a H5T_NATIVE_FLOAT type. If there is, and the data are stored as floats, then yeah I think I agree that that could make a big difference and run time. I’ll check it out.

-Aaron

**Comment by dagewa** [comment]
How are we doing with this one? @phyy-nx, you were going to check out the h5 float type?

**Comment by phyy-nx** [comment]
Hi, we used this branch during our experiment that concluded earlier this week. I've tested adding float but it's tricky.  Adding `custom_read<float>` works because `H5T_NATIVE_FLOAT` exists.  However, I need to switch which function I am using based on the numpy dtype, and while it's easy to check for a floating number type (`numpy.issubdtype(dataset.dtype, numpy.floating)`), determining if the dtype is float or double isn't as well specified.  

See https://docs.scipy.org/doc/numpy-1.13.0/reference/arrays.scalars.html

My best guess is the code would look like:
```
def dataset_as_flex(dataset, selection):
    if numpy.issubdtype(dataset.dtype, numpy.integer):
        return dataset_as_flex_int(dataset.id.id, selection)
    else:
        assert numpy.issubdtype(dataset.dtype, numpy.floating)
        if dataset.dtype in [numpy.half, numpy.single, numpy.double numpy.float_, numpy.float16]:
            return dataset_as_flex_float(dataset.id.id, selection)
        elif dataset.dtype in [numpy.longfloat numpy.float32, numpy.float64, numpy.float96, numpy.float128]:
            return dataset_as_flex_double(dataset.id.id, selection)
        else:
            assert False, "unknown floating data type (%s)"%str(dataset.dtype)
```

I don't really like it but I don't see another way.  I'll add this and merge the branch unless there's another suggestion :)

**Comment by phyy-nx** [comment]
Added the float version.  Would be cool if @jmp1985 could look as I needed to add a float version of ImageBuffer/Image as well.

**Comment by phyy-nx** [comment]
Ok, Ima merge it in the interest of getting this into production.  Thanks.

**Pull request merged**